File: 7Game/src/main/java/sevens/model/carddeck/Hand.java
package sevens.model.carddeck;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * The type Hand - represents a hand of playing cards.
 *
 * @author  Matthew Williams
 * @version 1.0
 * @since   2019-06-24
 */
public class Hand {

    private List<Card> hand;

    /**
     * Instantiates a new Hand.
     */
    public Hand() {
        hand = new ArrayList<>();
    }

    /**
     * Add card.
     *
     * @param card the card
     */
    public void addCard(Card card) {
        Objects.requireNonNull(card, "Invalid Card to Add");
        hand.add(card);
    }

    /**
     * Gets card.
     *
     * @param cardPosition the card position
     * @return the card
     */
    public Card getCard(int cardPosition) {
        checkValidPosition(cardPosition);
        return hand.get(cardPosition);
    }

    /**
     * Gets card.
     *
     * @param suit the suit of the card to retrieve
     * @param rank the rank of the card to retrieve
     * @return  the card if present in the users hand;
     *          null otherwise.
     */
    public Card getCard(int suit, int rank) {
        // check hand for card presence
        for (Card c : hand) {
            if ((c.getSuit() == suit) && (c.getRank() == rank)) {
                return c;
            }
        }
        // not present in hand
        return null;
    }

    /**
     * Gets the hand.
     *
     * @return the hand
     */
    public List<Card> getHand() {
        return hand;
    }

    /**
     * Remove card.
     *
     * @param card the card
     */
    public void removeCard(Card card) {
        hand.remove(card);
    }

    /**
     * Remove card.
     *
     * @param cardPosition the card position
     */
    public void removeCard(int cardPosition) {
        checkValidPosition(cardPosition);
        hand.remove(cardPosition);
    }

    /**
     * Sort hand by card rank.
     */
    public void sortByRank() {
        List<Card> newHand = new ArrayList<>();
        int minimalCardPosition;
        Card minimalCard;
        Card comparisonCard;
        while (hand.size() > 0) {
            minimalCardPosition = 0;
            minimalCard = hand.get(minimalCardPosition);
            for (int i = 1; i < hand.size(); i++) {
                comparisonCard = hand.get(i);
                if ((minimalCard.getRank() > comparisonCard.getRank()) ||
                        ((minimalCard.getRank() == comparisonCard.getRank()) && (minimalCard.getSuit() > comparisonCard.getSuit()))) {
                    // comparison card is lower in rank
                    minimalCardPosition = i;
                    minimalCard = comparisonCard;
                }
            }
            newHand.add(minimalCard);
            hand.remove(minimalCardPosition);
        }
        hand = newHand;
    }

    /**
     * Sort hand by card suit.
     */
    public void sortBySuit() {
        List<Card> newHand = new ArrayList<>();
        int minimalCardPosition;
        Card minimalCard;
        Card comparisonCard;
        while (hand.size() > 0) {
            minimalCardPosition = 0;
            minimalCard = hand.get(minimalCardPosition);
            for (int i = 1; i < hand.size(); i++) {
                comparisonCard = hand.get(i);
                if ((minimalCard.getSuit() > comparisonCard.getSuit()) ||
                        ((minimalCard.getSuit() == comparisonCard.getSuit()) && (minimalCard.getRank() > comparisonCard.getRank()))) {
                    // comparison card is lower in suit
                    minimalCardPosition = i;
                    minimalCard = comparisonCard;
                }
            }
            newHand.add(minimalCard);
            hand.remove(minimalCardPosition);
        }
        hand = newHand;
    }

    /**
     * Gets the number of cards in the hand.
     *
     * @return the card count in the hand
     */
    public int getCardCount() {
        return hand.size();
    }

    /**
     * Clear.
     */
    public void clear() {
        hand.clear();
    }

    private void checkValidPosition(int cardPosition) {
        if (cardPosition < 0 || cardPosition >= hand.size())
            throw new IllegalArgumentException("Invalid Position in Hand");
    }

}

File: 7Game/src/Hand.cpp
#include "Hand.h"
#include <algorithm>
#include <stdexcept>

Hand::Hand() {}

void Hand::addCard(const Card &card) { hand.push_back(card); }

Card Hand::getCard(int cardPosition) const {
  checkValidPosition(cardPosition);
  return hand[cardPosition];
}

std::optional<Card> Hand::getCard(int suit, int rank) const {
  for (const auto &c : hand) {
    if (c.getSuit() == suit && c.getRank() == rank) {
      return c;
    }
  }
  return std::nullopt;
}

const std::vector<Card> &Hand::getHand() const { return hand; }

void Hand::removeCard(const Card &card) {
  auto it = std::find_if(hand.begin(), hand.end(), [&](const Card &c) {
    return c.getSuit() == card.getSuit() && c.getRank() == card.getRank();
  });
  if (it != hand.end()) {
    hand.erase(it);
  }
}

void Hand::removeCardAt(int cardPosition) {
  checkValidPosition(cardPosition);
  hand.erase(hand.begin() + cardPosition);
}

void Hand::sortByRank() {
  std::sort(hand.begin(), hand.end(), [](const Card &a, const Card &b) {
    if (a.getRank() != b.getRank()) {
      return a.getRank() < b.getRank();
    }
    return a.getSuit() < b.getSuit();
  });
}

void Hand::sortBySuit() {
  std::sort(hand.begin(), hand.end(), [](const Card &a, const Card &b) {
    if (a.getSuit() != b.getSuit()) {
      return a.getSuit() < b.getSuit();
    }
    return a.getRank() < b.getRank();
  });
}

int Hand::getCardCount() const { return (int)hand.size(); }

void Hand::clear() { hand.clear(); }

void Hand::checkValidPosition(int cardPosition) const {
  if (cardPosition < 0 || cardPosition >= (int)hand.size()) {
    throw std::out_of_range("Invalid Position in Hand");
  }
}
